File: agent/runtime/src/brief.ts
// The turn brief is the contract between the conductor and this runtime. This file is its
// single source of truth: the Go mirror in internal/agent/conductor/brief.go and the golden
// fixture in testdata/brief.example.json both follow it, not the other way round.

import { z } from "zod";

import { MaxSkillNameLen, SkillNameRE } from "./skills.js";

/** BriefEnv holds base64(JSON) of one turn brief on the task spec. */
export const BriefEnv = "PODIUM_AGENT_TURN";

/**
 * MaxArgStrlen is Linux's cap on ONE environment string — MAX_ARG_STRLEN, which the kernel
 * fixes at 32 * PAGE_SIZE, so 128 KiB wherever the page is 4 KiB. Past it a container cannot
 * exec at all, which is why the brief's cap is not a matter of taste: the failure happens
 * before this file ever runs, so nothing reports it. Bisected against a real container, the
 * largest `PODIUM_AGENT_TURN` a `/bin/sh` will exec with is 131053 bytes — that plus
 * `"PODIUM_AGENT_TURN=".length` plus the NUL is exactly 131072.
 */
export const MaxArgStrlen = 32 * 4096;

/**
 * MaxBriefBytes caps the *encoded* brief. The conductor is what truncates a transcript that
 * does not fit — oldest entries first, `transcript_truncated: true` — and the runtime only
 * refuses, because a runtime that silently dropped context would answer the wrong question.
 *
 * It mirrors internal/agent/conductor.MaxBriefBytes and has to: a runtime that accepted more
 * than the conductor emits, or less, is a bug of its own. That file carries the reasoning for
 * the number — a quarter of MaxArgStrlen left spare, and still 72 KiB of JSON once base64 is
 * paid for.
 */
export const MaxBriefBytes = 96 * 1024;

/** ExitBriefInvalid is the runtime's exit code for anything wrong with the brief. */
export const ExitBriefInvalid = 2;

/** ReservedRepoName is the workspace directory Podium itself owns. */
export const ReservedRepoName = ".podium";

/**
 * McpNameRE constrains an MCP server's name. It is the conductor's own rule
 * (internal/agent/mcp.NameRE) and it is re-stated rather than trusted, because the name is
 * written into the harness config as a key and as a tool prefix — `linear` is what
 * `mcp__linear__*` comes from.
 */
export const McpNameRE = /^[a-z][a-z0-9-]{0,31}$/;

/** RepoNameRE constrains repos[].name: it becomes a directory under /workspace. */
export const RepoNameRE = /^[a-z0-9][a-z0-9._-]{0,63}$/;

/**
 * EffortLevels are the levels a model may be asked to think at. The conductor has already
 * refused one the chosen model does not accept, so anything that arrives here is passed
 * straight through to the harness.
 */
export const EffortLevels = ["low", "medium", "high", "xhigh", "max"] as const;

/** BriefError is every way a brief can be unusable. It always means exit 2. */
export class BriefError extends Error {
  readonly exitCode = ExitBriefInvalid;

  constructor(message: string) {
    super(message);
    this.name = "BriefError";
  }
}

const transcriptEntrySchema = z.strictObject({
  role: z.enum(["user", "assistant"]),
  author: z.string(),
  ts: z.string(),
  text: z.string(),
});

const repoSchema = z.strictObject({
  name: z
    .string()
    .regex(RepoNameRE, `must match ${RepoNameRE.source}`)
    // Unreachable while the pattern above rejects a leading dot, and kept anyway: the name
    // becomes a directory next to Podium's own, and that must never be an accident.
    .refine((n) => n !== ReservedRepoName, `may not be ${ReservedRepoName}`),
  url: z.string().min(1),
  default_branch: z.string().min(1),
});

// One Agent Skill the turn may use. It carries a name and a digest and never the bytes: the
// bundle rides in its own environment variable, exactly as a credential does. See skills.ts
// for why, and internal/agent/conductor/brief.go for the other half of the contract.
const skillRefSchema = z.strictObject({
  name: z.string().regex(SkillNameRE, `must match ${SkillNameRE.source}`).max(MaxSkillNameLen),
  // Hex, lower case, 64 characters: sha256 of the bundle document.
  sha256: z.string().regex(/^[0-9a-f]{64}$/, "must be a hex sha256 digest"),
  bundle_env: z.string().min(1),
});

// One playbook a host turn may delegate to. It says what the playbook can REACH rather than
// describing it, because that is what a model choosing between playbooks needs: which one
// has a Docker daemon, which one has the repository checked out.
const delegablePlaybookSchema = z.strictObject({
  name: z.string().min(1),
  summary: z.string().optional(),
  docker: z.boolean().optional(),
  browser: z.boolean().optional(),
  repos: z.array(z.string()).optional(),
});

// One MCP server the turn may use. An address and, when the server needs one, the NAME of
// the environment variable its bearer token arrives in — never the token. How it is
// presented is not carried: the MCP authorization specification says `Authorization: Bearer`
// and that is what writeConfig writes, so there is nothing per-server to describe.
const mcpServerSchema = z.strictObject({
  name: z.string().regex(McpNameRE, `must match ${McpNameRE.source}`),
  url: z.string().min(1),
  token_env: z.string().min(1).optional(),
});

const briefSchema = z.strictObject({
  version: z.literal(1),
  session_id: z.string().min(1),
  turn_id: z.string().min(1),
  source: z.strictObject({
    kind: z.enum(["slack", "linear", "chat"]),
    ref: z.string().min(1),
    url: z.string().optional(),
  }),
  profile: z.strictObject({
    name: z.string().min(1),
    display_name: z.string().min(1),
    system_prompt: z.string(),
    model: z.string().min(1),
    // Absent means the model's own default effort, which is the provider's choice.
    effort: z.enum(EffortLevels).optional(),
  }),
  playbook: z.strictObject({
    name: z.string().min(1),
    system_prompt: z.string(),
    allowed_tools: z.array(z.string()),
    // Absent means NO CAP, which is what the assistant runs with: it answers a conversation
    // and delegates, so the thing worth bounding is the container it starts and not the
    // relay that started it. A cap that killed a conversation mid-answer produced "I ran out
    // of turns" — a failure message for a turn that had not failed.
    //
    // A task always carries one: it runs unattended on a node, where nobody is watching a
    // cursor and the only other bound is the playbook's timeout.
    max_turns: z.number().int().positive().optional(),
    // The Agent Skills this turn may use. Absent means none — and the harness is handed a
    // permission map that denies every skill either way, so "no skills" is a decision
    // this runtime states rather than one it leaves to a default.
    skills: z.array(skillRefSchema).optional(),
    // The MCP servers this turn may use. Absent means none, and none is what the harness
    // gets: writeConfig writes one entry per server here beside the ones the conductor
    // wires up itself, and no others.
    mcp_servers: z.array(mcpServerSchema).optional(),
    // Absent means the turn cannot wait for a human: the default, and what every brief
    // before interactive playbooks described.
    interactive: z.boolean().optional(),
  }),
  transcript: z.array(transcriptEntrySchema),
  transcript_truncated: z.boolean(),
  instruction: z.string().min(1),
  repos: z.array(repoSchema).optional(),
  memory: z
    .strictObject({
      mcp_url: z.string().min(1),
      api_key_env: z.string().min(1),
    })
    .optional(),
  // The headless browser running beside this turn, when the playbook asked for one. An
  // address and nothing else: it is reached over the task's own private network, so unlike
  // memory there is no credential to name.
  browser: z
    .strictObject({
      cdp_url: z.string().min(1),
    })
    .optional(),
  // Which model API serves this turn. It is REQUIRED and it names no vendor in this file:
  // `id` is whatever the harness calls that provider, and together with profile.model it is
  // the whole of "what runs this turn".
  //
  // It carries the NAME of a secret and never a value, exactly as memory does — a brief is
  // an environment variable on a task spec, readable by anything that can read the spec.
  provider: z.strictObject({
    // id is the harness's provider id: "anthropic", "xai". Paired with profile.model it
    // becomes the harness's `provider/model`.
    id: z.string().min(1),
    // api_key_env names the environment variable the conductor put the credential in.
    api_key_env: z.string().min(1),
    // base_url overrides where that provider is reached — an egress proxy, or a test seam.
    // Absent means the harness's own default for the provider.
    base_url: z.string().optional(),
  }),
  // How this turn reaches a container, present only on a HOST turn: the conductor's own
  // address, the variable holding the token that authorises this turn to use it, and the
  // playbooks it may ask for.
  //
  // A TASK's brief never carries this, which is what stops a delegated task delegating
  // again. And `playbooks` is the allow-list as well as the menu: the conductor refuses a
  // name that is not on it, so the two cannot drift apart.
  delegation: z
    .strictObject({
      url: z.string().min(1),
      token_env: z.string().min(1),
      playbooks: z.array(delegablePlaybookSchema).min(1),
    })
    .optional(),
  // Where this turn is running. Absent means a task container, which is what every brief
  // before host turns described; "host" means a child process of the conductor, with no
  // container around it, no repository checked out and nothing collecting its files. The
  // prompt says different things in the two cases and must not have to guess which it is.
  runs_on: z.enum(["task", "host"]).optional(),
});

export type TurnBrief = z.infer<typeof briefSchema>;
export type TranscriptEntry = z.infer<typeof transcriptEntrySchema>;
export type RepoRef = z.infer<typeof repoSchema>;
export type SkillRef = z.infer<typeof skillRefSchema>;
export type McpServerRef = z.infer<typeof mcpServerSchema>;
export type SourceKind = TurnBrief["source"]["kind"];
export type ProviderRef = TurnBrief["provider"];
export type DelegationRef = NonNullable<TurnBrief["delegation"]>;

/** onHost reports whether this turn is a child of the conductor rather than a container. */
export function onHost(brief: TurnBrief): boolean {
  return brief.runs_on === "host";
}

export type DelegablePlaybook = z.infer<typeof delegablePlaybookSchema>;

/**
 * decodeBrief reads, decodes and validates the brief. Every failure is a BriefError, and
 * every BriefError is exit 2: a misspelt field must not silently change a turn, which is
 * why the schema is strict about unknown keys the way pkg/spec is about unknown YAML.
 */
export function decodeBrief(env: NodeJS.ProcessEnv): TurnBrief {
  const encoded = env[BriefEnv];
  if (encoded === undefined || encoded === "") {
    throw new BriefError(`${BriefEnv} is not set`);
  }

  const size = Buffer.byteLength(encoded, "utf8");
  if (size > MaxBriefBytes) {
    throw new BriefError(`turn brief is ${size} bytes; the limit is ${MaxBriefBytes}`);
  }

  const text = Buffer.from(encoded, "base64").toString("utf8");
  let raw: unknown;
  try {
    raw = JSON.parse(text) as unknown;
  } catch (err) {
    throw new BriefError(`${BriefEnv} is not base64-encoded JSON: ${messageOf(err)}`);
  }

  // Checked before the schema so the message can name what arrived: a version this runtime
  // does not speak is the one brief failure an operator can act on without reading a schema.
  const version = (raw as { version?: unknown } | null)?.version;
  if (version !== 1) {
    throw new BriefError(
      `turn brief version is ${JSON.stringify(version) ?? "absent"}; this runtime speaks version 1`,
    );
  }

  const parsed = briefSchema.safeParse(raw);
  if (!parsed.success) {
    throw new BriefError(`turn brief does not validate: ${formatIssues(parsed.error)}`);
  }
  return parsed.data;
}

function formatIssues(error: z.ZodError): string {
  return error.issues
    .map((issue) => {
      const path = issue.path.map((p) => String(p)).join(".");
      return path === "" ? issue.message : `${path}: ${issue.message}`;
    })
    .join("; ");
}

function messageOf(err: unknown): string {
  return err instanceof Error ? err.message : String(err);
}

File: internal/agent/conductor/brief.go
package conductor

import (
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"time"
)

// BriefVersion is the only version this conductor writes and the runtime reads.
const BriefVersion = 1

// BriefEnv is the environment variable one turn brief travels in, base64(JSON), on the task
// spec.
const BriefEnv = "PODIUM_AGENT_TURN"

// maxArgStrlen is Linux's cap on ONE environment string — MAX_ARG_STRLEN, which the kernel
// fixes at 32 * PAGE_SIZE. That is 128 KiB wherever the page is 4 KiB, which is every kernel
// Podium runs on; a bigger page only raises it, and nothing configures it down. So this is a
// floor rather than an estimate, and it is a floor worth respecting: past it the container
// cannot exec at all, which is the worst failure available here, because it happens before
// the runtime's entrypoint. Nothing reports it, the transcript is empty, and the task simply
// exits. Bisected against a real container, `NAME=value` and its NUL together:
//
//	$ docker run --rm -e "PODIUM_AGENT_TURN=$(python3 -c "print('x'*131053)")" alpine:3 /bin/sh -c 'printf ok'
//	ok
//	$ docker run --rm -e "PODIUM_AGENT_TURN=$(python3 -c "print('x'*131054)")" alpine:3 /bin/sh -c 'printf ok'
//	exec /bin/sh: argument list too long
//
// 131053 + len("PODIUM_AGENT_TURN=") + 1 for the NUL is exactly 131072.
//
// The cap is per string and not per environment: a brief at MaxBriefBytes beside eight skill
// bundles at skills.MaxEncodedBytes — 608 KiB of environment — execs. What bounds the whole
// of argv plus environ is a different and much larger limit, a quarter of RLIMIT_STACK, so
// about 2 MiB at the usual 8 MiB; docs/agent.md carries that measurement and nothing here
// budgets against it. TestTheBriefCapLeavesRoomToExec is what pins this one.
const maxArgStrlen = 128 << 10

// MaxBriefBytes caps the *encoded* brief — the base64, as it sits in the env var. The
// runtime refuses anything larger (exit 2) and never truncates; truncating is this side's
// job, because only the conductor knows which part of the context is oldest.
//
// It has to sit under maxArgStrlen with room to spare, or a brief that passes validation
// makes a task that cannot start. 96 KiB leaves a quarter of the limit unused, and it is not
// the smaller number it could be for one reason: a brief holds 3/4 of its cap as JSON, so
// 96 KiB is 72 KiB of document, and a 32 KiB chat message (api.maxChatMessageBytes) has to
// stay answerable beside two system prompts and some history.
const MaxBriefBytes = 96 * 1024

// The source kinds the runtime's schema accepts. Nothing else validates.
const (
	SourceSlack  = "slack"
	SourceLinear = "linear"
	SourceChat   = "chat"
)

// Brief mirrors agent/runtime/src/brief.ts field for field. That file is the single source
// of truth and its schema is strict at every level: an unknown key anywhere is exit 2, and
// an optional field must be *absent* rather than null. So every optional member below is a
// pointer or a slice with omitempty, and every required one is always emitted.
//
// The field order is the schema's order and the encoding is compact, because
// examples/agent/brief.sh renders the same document with `jq -cn` and a test compares bytes.
type Brief struct {
	Version             int           `json:"version"`
	SessionID           string        `json:"session_id"`
	TurnID              string        `json:"turn_id"`
	Source              BriefSource   `json:"source"`
	Profile             BriefProfile  `json:"profile"`
	Playbook            BriefPlaybook `json:"playbook"`
	Transcript          []BriefEntry  `json:"transcript"`
	TranscriptTruncated bool          `json:"transcript_truncated"`
	Instruction         string        `json:"instruction"`
	Repos               []BriefRepo   `json:"repos,omitempty"`
	// Git is who this turn's commits are by. Absent means the runtime's own fallback, which
	// is an identity no GitHub account owns — see profiles.GitPersona.
	Git *BriefGit `json:"git,omitempty"`
	// GitCredentials is where a turn redeems its minting capability for a GitHub token.
	// Absent means this conductor has no GitHub App, or this playbook has no repositories
	// to mint for, and the older path — a playbook naming its own token secret — applies.
	GitCredentials *BriefGitCredentials `json:"git_credentials,omitempty"`
	Memory         *BriefMemory         `json:"memory,omitempty"`
	Browser        *BriefBrowser        `json:"browser,omitempty"`
	Provider       *BriefProvider       `json:"provider,omitempty"`
	// Delegation is present only on a HOST turn's brief. A task's brief never carries one,
	// which is what stops a delegated task delegating again.
	Delegation *BriefDelegation `json:"delegation,omitempty"`
	// RunsOn is RunsOnHost for a turn running as a child of the conductor and absent for
	// one running in a task container. The runtime cannot infer it and must not guess: the
	// prompt tells a container turn it has a disposable filesystem and may leave files to
	// be attached, and both of those are false on a host.
	RunsOn string `json:"runs_on,omitempty"`
}

// RunsOnHost is Brief.RunsOn for a turn the conductor runs itself. The absent value means a
// task container, which is what every brief before host turns existed described.
const RunsOnHost = "host"

// BriefSource is where the turn came from and how a human reaches the conversation.
type BriefSource struct {
	Kind string `json:"kind"`
	Ref  string `json:"ref"`
	URL  string `json:"url,omitempty"`
}

// BriefProfile is the bot's identity for this turn, and the model it runs on.
type BriefProfile struct {
	Name         string `json:"name"`
	DisplayName  string `json:"display_name"`
	SystemPrompt string `json:"system_prompt"`
	Model        string `json:"model"`
	// Effort is the resolved reasoning effort, absent for the model's own default.
	Effort string `json:"effort,omitempty"`
}

// BriefProvider is which model API serves this turn. It is always emitted: `id` and
// profile.model together are the whole of what the harness is pointed at, and a turn with
// nowhere to send its requests cannot run.
//
// Nothing here names a vendor in Go: `id` is whatever the harness calls that provider, so
// adding a third is a catalogue entry rather than a change to this struct.
//
// APIKeyEnv is the NAME of the variable the credential lands in, never a value. A brief is
// an environment variable on a task spec and is visible to anything that can read the spec,
// exactly as memory.api_key_env is.
type BriefProvider struct {
	ID        string `json:"id"`
	APIKeyEnv string `json:"api_key_env"`
	// BaseURL overrides where the provider is reached — an egress proxy, or a test seam.
	// Empty means the harness's own default for that provider.
	BaseURL string `json:"base_url,omitempty"`
}

// BriefPlaybook is the job the turn is doing.
type BriefPlaybook struct {
	Name         string   `json:"name"`
	SystemPrompt string   `json:"system_prompt"`
	AllowedTools []string `json:"allowed_tools"`
	// MaxTurns caps the turn's steps. ZERO means no cap and is omitted from the document,
	// which is what the assistant runs with: the thing worth bounding is the container it
	// starts, not the relay that started it. A task always has one.
	MaxTurns int `json:"max_turns,omitempty"`
	// Skills is the Agent Skills this turn may use. Absent means none, and the runtime
	// writes a permission map that denies every skill either way.
	Skills []BriefSkill `json:"skills,omitempty"`
	// MCPServers is the MCP servers this turn may use. Absent means none, and the runtime
	// writes the harness one entry per server here and no others — so a playbook that names
	// none has no MCP tools beyond the ones the conductor wires up itself.
	MCPServers []BriefMCPServer `json:"mcp_servers,omitempty"`
	// Interactive is true when this turn may ask a human a question and wait. Absent
	// (false) is the default: the runtime has no ask tool and the prompt says so.
	Interactive bool `json:"interactive,omitempty"`
}

// BriefMCPServer points the runtime at one MCP server. It carries an address and the NAME of
// a credential, never a value — the same split memory and the model providers follow, for the
// same reason: the brief is an environment variable on a task spec and is readable by
// anything that can read the spec.
//
// How the credential is presented is not in here. `Authorization: Bearer <token>` is what
// the MCP authorization specification says and what the runtime writes; there is nothing
// per-server to carry.
type BriefMCPServer struct {
	// Name is the server's registered name, and the prefix the harness gives its tools.
	Name string `json:"name"`
	// URL is the server's endpoint.
	URL string `json:"url"`
	// TokenEnv names the environment variable holding the bearer token. Absent for a server
	// registered without one, which reaches its turns unauthenticated.
	TokenEnv string `json:"token_env,omitempty"`
}

// BriefSkill points the runtime at one Agent Skill bundle. It carries a NAME and a DIGEST,
// never the bytes: the bundle rides in its own environment variable, exactly as a
// credential does, and for the same reason — the brief is a document an operator reads, and
// a quarter of a megabyte of base64 in the middle of it is not readable.
//
// Keeping them apart also keeps the two caps apart. A brief that does not fit is truncated
// by dropping the oldest transcript entries; a bundle in the brief would therefore buy
// itself room by silently deleting the conversation, which is the wrong trade to make on a
// human's behalf.
type BriefSkill struct {
	// Name is the skill's name and the directory it is unpacked into.
	Name string `json:"name"`
	// SHA256 is the hex digest of the bundle document the runtime must verify before it
	// writes anything.
	SHA256 string `json:"sha256"`
	// BundleEnv names the environment variable holding base64(gzip(document)).
	BundleEnv string `json:"bundle_env"`
}

// BriefEntry is one utterance of the conversation so far.
type BriefEntry struct {
	Role   string `json:"role"`
	Author string `json:"author"`
	TS     string `json:"ts"`
	Text   string `json:"text"`
}

// BriefRepo is a repository the runtime clones into /workspace.
type BriefRepo struct {
	Name          string `json:"name"`
	URL           string `json:"url"`
	DefaultBranch string `json:"default_branch"`
}

// BriefGit is the user.name and user.email the runtime writes into every clone.
type BriefGit struct {
	Name  string `json:"name"`
	Email string `json:"email"`
}

// BriefGitCredentials points the runtime at this conductor's minting endpoint.
//
// It carries an address and the NAME of the variable the capability arrives in, exactly as
// BriefMemory does and for the same reason: a brief is an environment variable on a task
// spec, so it may name a credential and must never carry one.
type BriefGitCredentials struct {
	URL      string `json:"url"`
	TokenEnv string `json:"token_env"`
}

// BriefMemory points the runtime at Hindsight. Step 19 fills it; nothing here does.
type BriefMemory struct {
	MCPURL    string `json:"mcp_url"`
	APIKeyEnv string `json:"api_key_env"`
}

// BriefDelegation is how a host turn reaches a container: the conductor's own address, the
// variable holding the token that authorises this turn to use it, and the playbooks it may
// ask for.
//
// The playbook list is the ALLOW-LIST as well as the menu. It is built from the same profile
// snapshot as the token's grant, so what the model was shown and what the conductor will
// accept cannot drift apart between the fork and the call — a name that is not here is
// refused rather than resolved.
//
// TokenEnv is the NAME of a variable, never a value, exactly as memory.api_key_env is: a
// brief is a document, and one that carried a bearer token would be a document that must
// never be logged.
type BriefDelegation struct {
	URL       string              `json:"url"`
	TokenEnv  string              `json:"token_env"`
	Playbooks []DelegablePlaybook `json:"playbooks"`
}

// BriefBrowser points the runtime at the headless Chrome running beside this turn. It
// carries an address and nothing else: the browser is reached over the task's own private
// network, so there is no credential to name.
type BriefBrowser struct {
	// CDPURL is the DevTools endpoint of the browser sidecar.
	CDPURL string `json:"cdp_url"`
}

// Roles a transcript entry may carry.
const (
	RoleUser      = "user"
	RoleAssistant = "assistant"
)

// BriefTimestamp is how a transcript entry's ts is rendered. The schema only asks for a
// string; RFC 3339 in UTC is what the golden fixture uses and what a model can read.
func BriefTimestamp(t time.Time) string {
	return t.UTC().Format(time.RFC3339)
}

// ErrBriefTooLarge is what Encode returns when even an empty transcript does not fit. It
// means the prompts, the instruction or the repo list are themselves over the cap, and no
// amount of dropping history will help.
var ErrBriefTooLarge = errors.New("turn brief does not fit even with an empty transcript")

// Encode renders the brief for PODIUM_AGENT_TURN, dropping the oldest transcript entries
// until the encoded form fits MaxBriefBytes and setting TranscriptTruncated when it dropped
// any. The brief is modified in place so the caller can see what was actually sent.
func (b *Brief) Encode() (string, error) {
	b.normalise()
	for {
		encoded, err := b.encodeOnce()
		if err != nil {
			return "", err
		}
		if len(encoded) <= MaxBriefBytes {
			return encoded, nil
		}
		if len(b.Transcript) == 0 {
			return "", fmt.Errorf("%w: %d bytes encoded, the limit is %d",
				ErrBriefTooLarge, len(encoded), MaxBriefBytes)
		}
		// Oldest first: the end of a conversation is what the turn is answering.
		b.Transcript = b.Transcript[1:]
		b.TranscriptTruncated = true
	}
}

// normalise fills in what the schema requires to be present. A nil slice would marshal as
// null, and the runtime's schema wants [].
func (b *Brief) normalise() {
	if b.Version == 0 {
		b.Version = BriefVersion
	}
	if b.Transcript == nil {
		b.Transcript = []BriefEntry{}
	}
	if b.Playbook.AllowedTools == nil {
		b.Playbook.AllowedTools = []string{}
	}
}

// encodeOnce is the byte-exact encoder: compact JSON with HTML escaping off, because `jq`
// does not escape `<`, `>` or `&` either and examples/agent/brief.sh has to agree with this
// function byte for byte. json.Marshal would escape them; json.Encoder can be told not to.
func (b *Brief) encodeOnce() (string, error) {
	var buf strings.Builder
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(b); err != nil {
		return "", fmt.Errorf("encode turn brief: %w", err)
	}
	return base64.StdEncoding.EncodeToString([]byte(strings.TrimSuffix(buf.String(), "\n"))), nil
}

// briefKindFor is the source.kind the runtime's schema will accept for a session's source.
// The schema allows slack, linear and chat and nothing else, and the test-only dev source
// presents itself as chat.
//
// It exists because a DELEGATED task used to be told it came from the web chat whatever
// asked for it. That was invisible while only chats could delegate; it stopped being
// invisible when a Slack thread could, because the runtime's prompt tells the model where
// its answer is going and it was naming the wrong place.
func briefKindFor(kind string) string {
	if kind == KindDev {
		return SourceChat
	}
	return kind
}
